src/main/java/de/tum/in/www1/artemis/service/scheduled/LearningGoalProgressScheduleService.java

```diff
@@ -0,0 +1,203 @@
+            logger.error("Exception while processing progress task for learning goal {} and user {}:", learningObjectId, userId, e);
+        }
+        finally {
+            scheduledTasks.remove(new LearningGoalProgressId(learningObject, learningObjectId, userId).hashCode());
+        }
+        long end = System.currentTimeMillis();
+        logger.info("Updating the learning goal progress for learning goal {} and user {} took {} ms.", learningObjectId, userId, end - start);
+    }
+
+    private void updateLearningGoalProgress(Long learningGoalId, Long userId) {
+        var user = userRepository.findById(userId).orElse(null);
+        var learningGoal = learningGoalRepository.findByIdWithExercisesAndLectureUnitsAndCompletions(learningGoalId).orElse(null);
+
+        if (user == null || learningGoal == null) {
+            // If the user or learning goal no longer exist, there is nothing to do
+            return;
+        }
+
+        var progress = learningGoalProgressRepository.findEagerByLearningGoalIdAndUserId(learningGoalId, userId).orElse(new LearningGoalProgress());
+
+        List<ILearningObject> learningObjects = new ArrayList<>();
+
+        List<LectureUnit> allLectureUnits = learningGoal.getLectureUnits().stream().filter(LectureUnit::isVisibleToStudents).toList();
+
+        List<LectureUnit> lectureUnits = allLectureUnits.stream().filter(lectureUnit -> !(lectureUnit instanceof ExerciseUnit)).toList();
+        List<Exercise> exercises = learningGoal.getExercises().stream().filter(Exercise::isVisibleToStudents).toList();
+
+        learningObjects.addAll(lectureUnits);
+        learningObjects.addAll(exercises);
+
+        progress.setLearningGoal(learningGoal);
+        progress.setUser(user);
+        progress.setProgress(RoundingUtil.roundScoreSpecifiedByCourseSettings(calculateProgress(learningObjects, user), learningGoal.getCourse()));
+        progress.setConfidence(RoundingUtil.roundScoreSpecifiedByCourseSettings(calculateConfidence(exercises, user), learningGoal.getCourse()));
+        learningGoalProgressRepository.save(progress);
+    }
+
+    private double calculateProgress(List<ILearningObject> learningObjects, User user) {
+        var completions = learningObjects.stream().map(ILearningObject -> hasUserCompleted(user, ILearningObject)).toList();
+        completions.forEach(completed -> logger.debug("{} completed {}", user.getLogin(), completed));
+        return completions.stream().mapToInt(completed -> completed ? 100 : 0).summaryStatistics().getAverage();
+    }
+
+    private double calculateConfidence(List<Exercise> exercises, User user) {
+        var studentScores = studentScoreRepository.findAllByExercisesAndUser(exercises, user);
+        var teamScores = teamScoreRepository.findAllByExercisesAndUser(exercises, user);
+        return Stream.concat(studentScores.stream(), teamScores.stream()).map(ParticipantScore::getLastRatedScore).mapToDouble(score -> score).summaryStatistics().getAverage();
```

Comment:
💡 I am unsure about using only the rated result here. Should we not use the always the `lastScore` instead of the `lastRatedScore`. Otherwise students can never progress in a learning goal once they have received their "official" result by training on an exercise like a quiz or programming exercise.

Reply:
Good point, I changed it to the last score for now. In the future we might use a more complicated formula to calculate the confidence score. 🙂